### PR TITLE
SDN-4163: Add ipsec state to telemetry

### PR DESCRIPTION
**Pull request by JoshSalomon**
This PR adds a metric for the ipsec mode. The metric created is a Gauge named `cluster_network_ipsec_state` with a constant value of 1 and 2 labels:
1. "mode" which lists the ipsec mode, one of `Disabled` `External` or `Full`
2. "is_legacy_api" which states the flavor of the API used - either `true`
   for legacy API (up to OCP V4.14) or `false` for the new API (OCP 4.15+)
    -- If the cluster uses non ovn-k network (a network that does not
        support ipsec) the value of this label is "N/A - ipsec not supported (non-OVN network)"

Then a Prometheus rule  is created to remove other labels and this rule `openshift:cluster_network_ipsec_state:sum` will be collected in the telemetry

Jira ticket https://issues.redhat.com/browse/MON-3707

### CODE REVIEW

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	} else if ipsecConfig.Mode == "" {
+		mode = string(operv1.IPsecModeFull)
+		flavor = "4.14"
+	} else {
+		mode = string(ipsecConfig.Mode)
+		flavor = "4.15"
+	}
+	if mode != recordedMode || flavor != recordedFlavor {
+		klog.V(5).Infof("IPsec Telemetry: (%s, %s), writing to %v", mode, flavor, ipsecStateGauge)
+		ipsecStateGauge.WithLabelValues(mode, flavor).Set(1)
```
Comment (jan--f) [inline pkg/util/ipsec/ipsec_telemetry.go:46]: I'm neither an expert on instrumentation nor sure whether the following is intentional.
`WithLabelValues` will create a new time series if you call it with a never seen before set of label values. So if the expectation here is to always have a single timeseries, you'll need to reset the vector before set. Otherwise you'll end up with multiple series if those label values change.

**Comment by jan--f** [comment]
It's not required but good practice to create a recording rule for telemetry metrics. Something like `sum by(mode, apiFlavor) cluster_network_ipsec_state` for example. This guards telemetry against later additions to the metrics label set, since those additional label dimensions would be aggregated away.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	recordedMode, recordedFlavor string = "", ""
+)
+
+// RegisterIpsecMetrics in the global registry
+func RegisterIpsecMetrics() {
+	registerMetrics.Do(func() {
+		ipsecStateGauge = metrics.NewGaugeVec(&metrics.GaugeOpts{
+			Name: "cluster_network_ipsec_state",
+			Help: "A metric with a constant '1' value labeled by the latest ipsecMode of the cluster",
+		}, []string{"mode", "apiFlavor"})
```
Comment (jan--f) [inline pkg/util/ipsec/ipsec_telemetry.go:25]:
```suggestion
		}, []string{"mode", "api_flavor"})
```
Prometheus has a strong preference for snake_case oder camelCase.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 1 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:47]: @jan--f Thanks - I am not a Prometheus expert and I wondered how to keep a single value (It was not critical,  since in most cases we will have 1 or 2 values, but what I want is always a single value).

**Comment by JoshSalomon** [comment]
@jan--f - This is my first experience with telemetry, so this can be a beginner question. I think I know how to create the prometheus rules, but I couldn't find in the documentation how it is connected to the telemetry - I see that in the file 0000_50_cluster-monitoring-operator_04-config.yaml I add the counter name, so should I do anything else after I create the rule?
(I thought of adding a directory (bindata/network/ipsec) and add a file rules.yaml in this directory - but how should I connect this file and the rule inside into the telemetry?)

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 1 above)
Comment (jan--f) [reply pkg/util/ipsec/ipsec_telemetry.go:47]: Right, this is unfortunately a discouraged but widespread practice. Prometheus best-practice would recommend to always emit all possible label dimensions and set 0 for the ones not observed.
That can however be expensive. For larger metric sets like this a custom collector is usually used, like in https://github.com/kubernetes/kubernetes/blob/20d0ab7ae808aaddb1556c3c38ca0607663c50ac/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L299

I suppose this metric is expected to rarely change in a running cluster? So the reset way is probably just fine here.

Comment (simonpasquier) [reply pkg/util/ipsec/ipsec_telemetry.go:47]: As @jan--f  said, the preferred option would be to expose all label values with a value of 1 for the mode being enabled:

```
cluster_network_ipsec_state{mode="disabled"} 0
cluster_network_ipsec_state{mode="external"} 1
cluster_network_ipsec_state{mode="full"} 0
```

Using Reset() should work fine though it's more error-prone. The other option if you want to keep only 1 series would be to use the custom collector pattern (https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics).

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	})
+}
+
+func UpdateIpsecTelemetry(ipsecConfig *operv1.IPsecConfig) {
+	var flavor, mode string
+	klog.V(5).Infof("IPsec Telemetry: %v", ipsecConfig)
+	RegisterIpsecMetrics()
+	if ipsecConfig == nil {
+		mode = string(operv1.IPsecModeDisabled)
+		flavor = "4.14"
```
Comment (simonpasquier) [inline pkg/util/ipsec/ipsec_telemetry.go:36]: What is the `flavor` label about? Is it the OCP version?

**Comment by JoshSalomon** [comment]
@dougbtv @kyrtapz - Can I get a review for this PR - it blocks my next PR of adding the counter to telemetry.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 3 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:36]: Yes - the API has changed between OCP 4.14 and OCP 4.15

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 1 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:47]: @jan--f - this metric is indeed expected to change very rarely in production clusters, once you enable ipsec you rarely disable it (assuming you had a business reason to enable it, it doesn't go away).

Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:47]: @simonpasquier - for a metric as simple as this which rarely changes, can I leave the current implementation? I believe the option you suggest will just add more lines of code for something really simple and will make the code cumbersome.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/util/ipsec/ipsec_telemetry.go:36]: I see no reason to have it as a label since the cluster version is already available in Telemeter.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 1 above)
Comment (simonpasquier) [reply pkg/util/ipsec/ipsec_telemetry.go:47]: I'm not sure that the code would be more complex with a custom collector. It's your call since you maintain that piece of code ;)

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 3 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:36]: I probably misled you in the previous response. There are 2 flavors to the API according to the version in which the API was invoked, but for backward compatibility we keep the API version during an upgrade until the API is invoked again, so you can find an instance of the old API (4.14) in newer clusters. The reason to keep the API flavor is in order to understand when we can drop the support for the old API flavor, which is non-trivial.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 4]
```diff
@@ -348,6 +349,7 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
 		}
 	}
 
+	UpdateIpsecTelemetry(&newOperConfig.Spec)
```
Comment (pperiyasamy) [inline pkg/controller/operconfig/operconfig_controller.go:352]: I think it would be good to introduce new field in `ReconcileOperConfig` struct to store IPsec mode and update ipsec metric when it changes to a new value.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+var (
+	registerMetrics sync.Once
+	ipsecStateGauge *metrics.GaugeVec
+
+	recordedMode, recordedFlavor string = "", ""
```
Comment (pperiyasamy) [inline pkg/util/ipsec/ipsec_telemetry.go:16]: As I said, this must be moved under `ReconcileOperConfig`

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+func UpdateIpsecTelemetry(ipsecConfig *operv1.IPsecConfig) {
+	var flavor, mode string
+	klog.V(5).Infof("IPsec Telemetry: %v", ipsecConfig)
+	RegisterIpsecMetrics()
+	if ipsecConfig == nil {
+		mode = string(operv1.IPsecModeDisabled)
+		flavor = "4.14"
+	} else if ipsecConfig.Mode == "" {
+		mode = string(operv1.IPsecModeFull)
```
Comment (pperiyasamy) [inline pkg/util/ipsec/ipsec_telemetry.go:38]: I'd reused existing method [getIPsecMode](https://github.com/openshift/cluster-network-operator/blob/master/pkg/network/ovn_kubernetes.go#L535) to retrieve the ipsec mode.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		flavor = "4.14"
+	} else {
+		mode = string(ipsecConfig.Mode)
+		flavor = "4.15"
```
Comment (pperiyasamy) [inline pkg/util/ipsec/ipsec_telemetry.go:42]: what's the significance of exposing flavor in the metric ? I'm not sure if it's good idea to hardcode these values with 4.15 or 4.14. Did you consider to use https://github.com/openshift/cluster-network-operator/blob/master/vendor/github.com/openshift/api/operator/v1/types.go#L123 ?

Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:42]: I am not sure I understand this - we have 2 versions of the API, and they changed between 4.14 and 4.15. I believe it is good to know if someone is still using the older version. I can call it "old" and "new", or any other agreed constant names, but if we collect telemetry, it is an important piece of information. I don't understand how the `Version` field can help me - since I need constants and not something that can change according to the version in which the code is executed.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 6 above)
Comment (pperiyasamy) [reply pkg/util/ipsec/ipsec_telemetry.go:38]: yes, you could just make `getIPsecMode` method as public and use it from here. I prefer to have new method for returning flavor.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 7 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:42]: This is an option - I am not sure if booleans are supported by Prometheus labels, but it can be "true" or "false" if not. I will do it.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		}, []string{"mode", "is_legacy_api"})
+		legacyregistry.MustRegister(ipsecStateGauge)
+	})
+}
+
+func UpdateIpsecTelemetry(state string, isLegacyAPI bool) {
+	klog.V(5).Infof("IPsec mode(prev): %s(%s), isLegacyAPI(prev): %v(%s)",
+		state, recordedMode.state, isLegacyAPI, recordedMode.legacyAPI)
+	isLegacyAPIStr := strconv.FormatBool(isLegacyAPI)
+	RegisterIpsecMetrics()
```
[the remark this answers is not in the document]
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:39]: registerMetrics is an instance of sync.Once - it runs only once.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	if state != recordedMode.state || isLegacyAPIStr != recordedMode.legacyAPI {
```
[the remark this answers is not in the document]
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:40]: It is not after upgrade - the API could be invoked from `ipsec: {}` to `ipsec.mode: "Full"`
from a functional perspective, it may not matter but I believe it is good for metrics purposes. Especially since there is no way back from the new API to the legacy one.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 10]
```diff
@@ -550,6 +552,25 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
...
+
+	newOVNKubeConfig := newOperConfigSpec.DefaultNetwork.OVNKubernetesConfig
+	klog.V(5).Infof("IPsec: calling UpdateIpsecTelemetry, %v", newOVNKubeConfig.IPsecConfig)
+	mode := string(network.GetIPsecMode(newOVNKubeConfig))
+	legacyAPI := network.IsIPsecLegacyAPI(newOVNKubeConfig)
+
+	ipsecTelemetry.UpdateIpsecTelemetry(mode, legacyAPI)
+
+	klog.V(5).Infof("IPsec: << UpdateIpsecTelemetry")
+	// Don't record telemetry when the IPsecConfig hasn't changed
```
[the remark this answers is not in the document]
Comment (JoshSalomon) [reply pkg/controller/operconfig/operconfig_controller.go:571]: will fix this (tomorrow)

File: pkg/controller/operconfig/operconfig_controller.go [hunk 11]
```diff
@@ -550,6 +552,25 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
 	return reconcile.Result{RequeueAfter: ResyncPeriod}, nil
 }
 
+func UpdateIpsecTelemetry(newOperConfigSpec *operv1.NetworkSpec) {
+	klog.V(5).Infof("IPsec: >> UpdateIpsecTelemetry new: %v", newOperConfigSpec)
```
[the remark this answers is not in the document]
Comment (JoshSalomon) [reply pkg/controller/operconfig/operconfig_controller.go:556]: will fix this (tomorrow)

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply pkg/util/ipsec/ipsec_telemetry.go:39]: ok, but this is still bit confusing calling register from update metrics method. 
Can you at least do it via module's init method if it can not be registered from operconfig controller ?

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+var (
+	registerMetrics sync.Once
+	ipsecStateGauge *metrics.GaugeVec
+
+	recordedMode ipsecAPIMode // initial values are empty strings
+)
+
+// RegisterIpsecMetrics in the global registry
+func RegisterIpsecMetrics() {
```
Comment (pperiyasamy) [inline pkg/util/ipsec/ipsec_telemetry.go:25]: Nit, replace Ipsec with IPsec, this is the pattern being used all over the places.

**Review by pperiyasamy — COMMENTED** [review]
Please discuss with @martinkennelly, @jcaamano about where to maintain current IPsec mode.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 12 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:25]: Regarding the registration - this is not something I invented - some modules (e.g. pkg/cmd/checkendpoints/controller/metrics.go) use this method, but some (e.g. vendor/github.com/grpc-ecosystem/go-grpc-prometheus/client.go) use the init module, I can move it. Good catch on the name (should be IPsec)

**Comment by JoshSalomon** [comment]
> Please discuss with @martinkennelly, @jcaamano about where to maintain current IPsec mode.

@martinkennelly @jcaamano - What I hold is not the ipsec mode is the ipsec *last recorded* mode which has more legal values than ipsec mode (it can be empty on cluster init), moreover it catches the situation where there is nothing to do (disabled mode) which is not caught in the operator code, since it requires no-op. So while in normal work I could have placed this code where the ipsec change state code is, and call the metric change on ipsec change, on startup this is not the case, and I see no reason to clutter the operator code more than it should be, for some metric optimization code which can be encapsulated in a small isolated piece of code.

File: bindata/network/network-metrics/ipsec-rules.yaml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+apiVersion: monitoring.coreos.com/v1
```
Comment (martinkennelly) [inline bindata/network/network-metrics/ipsec-rules.yaml:1]: is this for sdn + ovn-k? if its just for ovn-k and you want to expose this for every deployment model we have, id put this in `bindata/network/ovn-kubernetes/common/alert-rules.yaml`. If its not meant for managed services and just non-managed standalone clusters, you can adjust where to put it.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	recordedState     string
+	recordedLegacyAPI string
+)
+
+func init() {
+	ipsecStateGauge = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Name: "cluster_network_ipsec_state",
+		Help: "A metric with a constant '1' value labeled by the latest ipsecMode of the cluster",
+	}, []string{"mode", "is_legacy_api"})
+	legacyregistry.MustRegister(ipsecStateGauge)
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:23]: I am confused - how will the metrics be consumed from CNO? 
We dont expose any endpoint for Prometheus to scrape this metric AFAIK.

Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:23]: It will be consumed from the telemetry module - the Prometheus query "openshift:cluster_network_ipsec_state:sum" will be gathered by the telemetry agent

Comment (martinkennelly) [reply pkg/util/ipsec/ipsec_telemetry.go:23]: I mean, how will prometheus actually get the metrics info - i dont think we expose a server for prometheus agent to scrape the metrics. 
Did you see the metrics being exposed in OCP console with this PR?

Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:23]: Yep - 
image

Comment (martinkennelly) [reply pkg/util/ipsec/ipsec_telemetry.go:23]: I was wrong - its serving on 9104. Thanks!

File: bindata/network/network-metrics/ipsec-rules.yaml (hunk 13 above)
Comment (martinkennelly) [inline bindata/network/network-metrics/ipsec-rules.yaml:1]: i think move to manifests file and name `07-prometheus-rules.yaml`. WDYT?

This dir is for network metrics daemon.

File: pkg/controller/operconfig/operconfig_controller.go [hunk 15]
```diff
@@ -550,6 +552,19 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
 	return reconcile.Result{RequeueAfter: ResyncPeriod}, nil
 }
 
+func updateIPsecTelemetry(newOperConfigSpec *operv1.NetworkSpec) {
```
Comment (martinkennelly) [inline pkg/controller/operconfig/operconfig_controller.go:555]: `updateIPsecMetric(`

Theres no telemetry update here - just updating a metric. Gathering of telemetry based on this metric is outside the scope of this PR.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+var (
+	ipsecStateGauge *metrics.GaugeVec
+
+	recordedState     string
+	recordedLegacyAPI string
+)
+
+func init() {
+	ipsecStateGauge = metrics.NewGaugeVec(&metrics.GaugeOpts{
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:19]: please fill in fields `namespace`, `subsystem` and `name`. Its how the full name is usually constructed in OCP.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+		Name: "cluster_network_ipsec_state",
+		Help: "A metric with a constant '1' value labeled by the latest ipsecMode of the cluster",
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:21]: what about a description of the legacy api label?

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+var (
+	ipsecStateGauge *metrics.GaugeVec
+
+	recordedState     string
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:14]: i dont see the need for this. Just update the metric with the current state. Using these vars just made the code harder to read.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+	ipsecStateGauge = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Name: "cluster_network_ipsec_state",
+		Help: "A metric with a constant '1' value labeled by the latest ipsecMode of the cluster",
+	}, []string{"mode", "is_legacy_api"})
+	legacyregistry.MustRegister(ipsecStateGauge)
+	recordedState = ""
+	recordedLegacyAPI = ""
+}
+
+func UpdateIPsecTelemetry(state string, isLegacyAPI bool) {
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:28]: ` UpdateIPsecMetric`

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+	klog.V(5).Infof("IPsec mode(prev): %s(%s), isLegacyAPI(prev): %v(%s)",
+		state, recordedState, isLegacyAPI, recordedLegacyAPI)
+	isLegacyAPIStr := strconv.FormatBool(isLegacyAPI)
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:31]: i think you need another state besides true and false - if ipsec is disabled, then the third state should be `unknown` otherwise it looks strange defining is it legacy api or not when its not even configured. WDYT?

File: pkg/controller/operconfig/operconfig_controller.go [hunk 21]
```diff
@@ -550,6 +552,19 @@ func (r *ReconcileOperConfig) Reconcile(ctx context.Context, request reconcile.R
...
+	if newOperConfigSpec == nil || newOperConfigSpec.DefaultNetwork.OVNKubernetesConfig == nil {
```
Comment (martinkennelly) [inline pkg/controller/operconfig/operconfig_controller.go:556]: what if the user switches to a new CNI from ovn-k? the metrics will still report old values indicating ipsec is enabled. You should reset the metric for this scenario.

File: pkg/util/ipsec/ipsec_telemetry.go [hunk 22]
```diff
@@ -0,0 +1,39 @@
+package ipsec
```
Comment (martinkennelly) [inline pkg/util/ipsec/ipsec_telemetry.go:1]: rename file `metrics.go`

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 18 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:14]: IIUC - this causes having more load on Prometheus - as Prometheus holds a time series it does not collapse identical values - this is just another value in the time series, and this is not what we want to keep even if we can collapse this in the telemetry collection. Actually, this part of the code was reviewed (with comments) and approved by the @jan--f from the telemetry team.
@jan--f does it make sense just to write the same value to Prometheus over and over again when all we need is the latest state and not a time series?

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 16 above)
Comment (JoshSalomon) [reply pkg/util/ipsec/ipsec_telemetry.go:19]: I see that namespace is hardly used and this metric is cluster level and not really namespaced (in meaning that it is limited to a specific namespace and can differ between namespaces) so I just used only subsystem and name (which is the common way to create metrics as far as I could find)

File: bindata/network/network-metrics/ipsec-rules.yaml (hunk 13 above)
Comment (JoshSalomon) [reply bindata/network/network-metrics/ipsec-rules.yaml]: The file was not applied during the installation in the manifests directory, I am not sure why. On the other hand, this is kind of network metric...

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 16 above)
Comment (martinkennelly) [reply pkg/util/ipsec/ipsec_telemetry.go]: Its cluster level but in order to easily inform the consumer where it came from, we populate the `namespace` field.

File: pkg/util/ipsec/ipsec_telemetry.go (hunk 18 above)
Comment (martinkennelly) [reply pkg/util/ipsec/ipsec_telemetry.go]: Its just updating what the prom server (presented by CNO here) presents to the prometheus scraper - so if the value doesnt change, its a no-op and the server will present the same data. It doesnt increase the load on the prometheus backend which store this data.

Comment (martinkennelly) [reply pkg/util/ipsec/ipsec_telemetry.go]: I think it just makes it easier to read this code without affecting performance much. It wont increase the load or create additional artifacts in the prometheus backend db (in the prometheus namespace)

File: bindata/cluster-network-operator/prometheus.yaml [hunk 23]
```diff
@@ -0,0 +1,18 @@
+##
+# This file is not applied correctly during cluster installation when located in
+# the monifests directory, this is the reason why it is located in the bindata directory.
```
Comment (martinkennelly) [inline bindata/cluster-network-operator/prometheus.yaml:3]: monifests -> manifests

**Review by martinkennelly — APPROVED** [review]

**Comment by kyrtapz** [comment]
@JoshSalomon the `e2e-gcp-sdn` job reports that CNO is crashing:
`level=error msg=Cluster operator network Degraded is True with ReconcileError: Panic detected: runtime error: invalid memory address or nil pointer dereference`
https://storage.googleapis.com/test-platform-results/pr-logs/pull/openshift_cluster-network-operator/2270/pull-ci-openshift-cluster-network-operator-master-e2e-gcp-sdn/1757339435860496384/build-log.txt

/test e2e-gcp-sdn
